pynest/examples/eprop_plasticity/eprop_supervised_classification_evidence-accumulation_bsshslm_2020.py

```diff
@@ -578,13 +601,115 @@ def get_weights(pop_pre, pop_post):
+
+    y_prediction = np.argmax(np.mean(readout_signal, axis=3), axis=0)
+    y_target = np.argmax(np.mean(target_signal, axis=3), axis=0)
+    accuracy = np.mean((y_target == y_prediction), axis=1)
+    errors = 1.0 - accuracy
+
+    results_dict["iteration"].extend(range(n_iter_start, n_iter_start + n_iter_interval))
+    results_dict["error"].extend(errors)
+    results_dict["loss"].extend(loss)
+    results_dict["label"].extend([phase_label] * len(loss))
+
+    return errors, loss
+
+
 # %% ###########################################################################################################
 # Simulate
 # ~~~~~~~~
-# We train the network by simulating for a set simulation time, determined by the number of iterations and the
-# batch size and the length of one sequence.
+# We train the network by simulating for a number of training iterations with the set learning rate. If early
+# stopping is turned on, we evaluate the network's performance on the validation set in regular intervals and,
+# if the error is below a certain threshold, we stop the training early. If the error is not below the
+# threshold, we continue training until the end of the set number of iterations. Finally, we evaluate the
+# network's performance on the test set.
+
+results_dict = {
+    "error": [],
+    "loss": [],
+    "iteration": [],
+    "label": [],
+}
 
-nest.Simulate(duration["sim"])
+
+def run(phase_label, n_iter_start, eta, phase_label_previous, n_iter_interval=1):
+    params_common_syn_eprop["optimizer"]["eta"] = eta
+    nest.SetDefaults("eprop_synapse_bsshslm_2020", params_common_syn_eprop)
+
+    nest.Simulate(duration["extension_sim"])
+    error_val = [0]
+    if n_iter_start > 0:
+        error_val, loss = evaluate(n_iter_start, n_iter_interval, phase_label_previous)
+
+    duration["sim"] = n_iter_interval * batch_size * duration["sequence"] - duration["extension_sim"]
+
+    nest.Simulate(duration["sim"])
+
+    n_iter_start_new = n_iter_start + n_iter_interval
+    return error_val, n_iter_start_new, phase_label
+
+
+n_iter_sim = 0
+nest.Simulate(duration["total_offset"])
+
+phase_label_previous = ""
+for k_iter in range(n_iter):
+    if do_early_stopping and k_iter % n_validate_every == 0:
+        error_val, n_iter_sim, phase_label_previous = run("validation", n_iter_sim, eta_test, phase_label_previous)
+
+        if k_iter > 0 and error_val < stop_crit:
+            errors_early_stop, n_iter_sim, phase_label_previous = run(
+                "early-stopping", n_iter_sim, eta_test, phase_label_previous
+            )
+            if np.mean(errors_early_stop) < stop_crit:
+                break
+
+    run_iter = min(n_iter - k_iter, n_validate_every)
+    _, n_iter_sim, phase_label_previous = run("training", n_iter_sim, eta_train, phase_label_previous)
+
+for _ in range(n_test):
+    _, n_iter_sim, phase_label_previous = run("test", n_iter_sim, eta_test, phase_label_previous)
+
+nest.Simulate(steps["extension_sim"])
```

Comment:
Could the loop structure be reorganized to make the phase separation clearer? Currently, the only indication is the string labels, but they don’t seem obvious enough. Would a reorganization like the following be more effective?

```python
def simulate_training_phase(n_iter_sim, eta_train, phase_label_previous):
    run_iter = min(n_iter - k_iter, n_validate_every)
    return run("training", n_iter_sim, eta_train, phase_label_previous)


def simulate_validation_phase(n_iter_sim, eta_test, phase_label_previous):
    return run("validation", n_iter_sim, eta_test, phase_label_previous)


def simulate_early_stopping_phase(n_iter_sim, eta_test, phase_label_previous):
    errors_early_stop, n_iter_sim, phase_label_previous = run("early-stopping", n_iter_sim, eta_test, phase_label_previous)
    if np.mean(errors_early_stop) < stop_crit:
        return True, n_iter_sim, phase_label_previous
    return False, n_iter_sim, phase_label_previous


def simulate_test_phase(n_test, n_iter_sim, eta_test, phase_label_previous):
    for _ in range(n_test):
        run("test", n_iter_sim, eta_test, phase_label_previous)


n_iter_sim = 0
nest.Simulate(duration["total_offset"])
phase_label_previous = ""

for k_iter in range(n_iter):
    # Validation phase and early stoping check
    if do_early_stopping and k_iter % n_validate_every == 0:
        error_val, n_iter_sim, phase_label_previous = simulate_validation_phase(n_iter_sim, eta_test, phase_label_previous)

        if k_iter > 0 and error_val < stop_crit:
            should_stop, n_iter_sim, phase_label_previous = simulate_early_stopping_phase(n_iter_sim, eta_test, phase_label_previous)
            if should_stop:
                print(f"Early stopping at iteration {k_iter}")
                break

    # Training phase
    n_iter_sim, phase_label_previous = simulate_training_phase(n_iter_sim, eta_train, phase_label_previous)

# Test phase
simulate_test_phase(n_test, n_iter_sim, eta_test, phase_label_previous)
```

Reply:
Nice idea! Please, see https://github.com/JesusEV/nest-simulator/pull/28/commits/4c13272ecca1e5d43b8501cbae3af5c9e3c1516d

Reply:
It looks great. The compartmentalization makes it look much cleaner and more organized. I'm sure this should make maintenance and modification easier. In particular I like the new `run_early_stopping` function, which isolates the core early stopping logic into few, easy-to-follow lines.